Updated desc_sroa to support flattening structures

ehsannas opened the pull request:
This doesn't cover cases where we have the following pattern:
```
         %31 = OpAccessChain %_ptr_UniformConstant_S %globalS %int_0 %int_0
         %32 = OpLoad %S %31
         %33 = OpCompositeExtract %_arr_type_2d_image_uint_2 %32 0
         %34 = OpCompositeExtract %type_2d_image %33 0
         %35 = OpCompositeExtract %type_2d_image %33 1
         %36 = OpCompositeExtract %_arr_T_uint_2 %32 1
         %37 = OpCompositeExtract %T %36 0
         %38 = OpCompositeExtract %_arr_type_sampler_uint_3 %37 0
         %39 = OpCompositeExtract %type_sampler %38 1
         %40 = OpCompositeExtract %T %36 1
         %41 = OpCompositeExtract %_arr_type_sampler_uint_3 %40 0
         %42 = OpCompositeExtract %type_sampler %41 2
         %43 = OpSampledImage %type_sampled_image %34 %39
         %44 = OpImageSampleImplicitLod %v4float %43 %14 None
```

This pattern could be seen if a global structure of resources is passed to a function.

ehsannas commented on source/opt/ir_context.h:
```diff
@@ -1061,7 +1068,9 @@ void IRContext::AddDebug1Inst(std::unique_ptr<Instruction>&& d) {
-      id_to_name_->insert({d->result_id(), d.get()});
+      // OpName and OpMemberName do not have result-ids. The target of the
+      // instruction is at InOperand index 0.
+      id_to_name_->insert({d->GetSingleWordInOperand(0), d.get()});
```
fixed an existing bug

ehsannas commented on source/opt/desc_sroa.cpp:
```diff
@@ -242,32 +274,85 @@ uint32_t DescriptorScalarReplacement::CreateReplacementVariable(
     Instruction* new_name_inst = new_name.get();
-    context()->AddDebug2Inst(std::move(new_name));
     get_def_use_mgr()->AnalyzeInstDefUse(new_name_inst);
+    names_to_add.push_back(std::move(new_name));
```
fixed an existing bug

s-perron commented on source/opt/desc_sroa.cpp:
```diff
@@ -242,32 +274,85 @@ uint32_t DescriptorScalarReplacement::CreateReplacementVariable(
+  }
+
+  // Structures consume M binding numbers where M is the sum of binding numbers
+  // used by their members.
```
No need to mention M in the comment, since it is never repeated.  Just say

> The number of bindings consumed by a structure is the sum of the bindings used by its members.

s-perron commented on source/opt/desc_sroa.h:
```diff
@@ -70,6 +70,15 @@ class DescriptorScalarReplacement : public Pass {
+  // 2- An array of size N consumes N*M binding numbers, where M is the number
+  // of bindings used by each array element.
+  // 3- A structure consumes M binding numbers where M is the sum of binding
+  // numbers used by its members.
```
Same as above.

s-perron commented on source/opt/ir_context.h:
```diff
@@ -357,6 +357,13 @@ class IRContext {
   GetNames(uint32_t id);
 
+  // Returns an OpMemberName instruction that targets |struct_type_id| at
+  // index |index|. Returns nullptr if such instruction does not exist.
```
replace "if such instruction does not exist" with "if no such instruction exists"

ehsannas commented:
I realized that the check I had added missed cbuffers (cbuffers would have been expanded). I've updated the code to check for `Offset` decoration on the variable type to determine if a struct is a buffer.

ehsannas commented:
the `CI-check-format` failure is not relevant to my change. Looks like `test/opt/loop_optimizations/loop_fission.cpp` has bad formatting.
